auth_with_external_browser_test.go

```diff
@@ -0,0 +1,205 @@
+		defer wg.Done()
+		expectedErrorMsg := "390191 (08004): The user you were trying to authenticate " +
+			"as differs from the user currently logged in at the IDP."
+
+		_, err := connectToSnowflake(cfg, "SELECT 1", false)
+		if err.Error() != expectedErrorMsg {
+			t.Errorf("Expected %v, but got %v", expectedErrorMsg, err)
+		}
+	}()
+	wg.Wait()
+}
+
+func TestClientStoreCredentials(t *testing.T) {
+	cfg := setupExternalBrowserTest(t)
+	cfg.ClientStoreTemporaryCredential = 1
+	cfg.ExternalBrowserTimeout = time.Duration(10000) * time.Millisecond
+
+	t.Run("Obtains the ID token from the server and saves it on the local storage", func(t *testing.T) {
+		cleanupBrowserProcesses()
+		var wg sync.WaitGroup
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			provideCredentials(externalBrowserType.Success, cfg.User, cfg.Password)
+		}()
+		go func() {
+			defer wg.Done()
+			conn, err := connectToSnowflake(cfg, "SELECT 1", true)
+			if err != nil {
+				t.Errorf("Connection failed: err %v", err)
+			}
+			defer conn.Close()
+		}()
+		wg.Wait()
+	})
+
+	t.Run("Verify validation of ID token if option enabled", func(t *testing.T) {
+		cleanupBrowserProcesses()
+		cfg.ClientStoreTemporaryCredential = 1
+		conn, _ := createConnection(getDbHandler(cfg))
+		_, err := conn.QueryContext(context.Background(), "SELECT 1")
+		if err != nil {
+			log.Fatalf("failed to run a query. err: %v", err)
+		}
+	})
+
+	t.Run("Verify validation of IDToken if option disabled", func(t *testing.T) {
+		cleanupBrowserProcesses()
+		cfg.ClientStoreTemporaryCredential = 0
+		tOut := "authentication timed out"
+		_, err := createConnection(getDbHandler(cfg))
+		if err.Error() != tOut {
+			t.Errorf("Expected %v, but got %v", tOut, err)
+		}
+	})
+}
+
+type Mode struct {
+	Success string
+	Fail    string
+	Timeout string
+}
+
+var externalBrowserType = Mode{
+	Success: "success",
+	Fail:    "fail",
+	Timeout: "timeout",
+}
+
+func cleanupBrowserProcesses() {
+	const cleanBrowserProcessesPath = "/externalbrowser/cleanBrowserProcesses.js"
+	_, err := exec.Command("node", cleanBrowserProcessesPath).Output()
+	if err != nil {
+		log.Fatalf("failed to execute command: %v", err)
+	}
+}
+
+func provideCredentials(mode string, user string, password string) {
+	const provideBrowserCredentialsPath = "/externalbrowser/provideBrowserCredentials.js"
+	_, err := exec.Command("node", provideBrowserCredentialsPath, mode, user, password).Output()
+	if err != nil {
+		log.Fatalf("failed to execute command: %v", err)
+	}
+}
+
+func connectToSnowflake(cfg *Config, query string, exceptionHandler bool) (rows *sql.Rows, err error) {
+	parseFlags()
+	dsn, err := DSN(cfg)
+	if err != nil {
+		log.Fatalf("failed to create DSN from Config: %v, err: %v", cfg, err)
+	}
+	rows, err = executeQuery(query, dsn)
+	if exceptionHandler && err != nil {
+		log.Fatalf("failed to run a query. %v, err: %v", rows, err)
+	} else if err != nil {
+		return rows, err
+	}
+	defer rows.Close()
+	var v int
+	for rows.Next() {
```

Comment:
It would be better to use `assertTrueF(t, rows.Next())`. If no row is returned, we never enter the assertions below.

Reply:
I didn't pass *t.testing object to the function so instead I can propose something like this:

Currently in line 176 we are handling error if any problems with rows occurred. 
In 186 line if rows.next() we were testing data against correctness.

Now, I will add another condition "if !rows.Next()" then I will inform user "There were no results for query: %v \n", query).

which I think is still fine, no error here, just no data for specific query.